### PR TITLE
Upgrade from Chromium 103 to Chromium 104.

### PR DESCRIPTION
**Pull request by mkarolin**
Fixes https://github.com/brave/brave-browser/issues/23092
Fixes https://github.com/brave/brave-browser/issues/22922

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/extensions/api/brave_rewards_api.cc
Comment (mkarolin) [inline browser/extensions/api/brave_rewards_api.cc:941]: @zenparsing: Is it ok that we are casting from int64_t to int here?

Comment (cdesouza-chromium) [reply browser/extensions/api/brave_rewards_api.cc:941]: Possibly we should use `base::Int64ToValue` there, and avoid the conversion altogether.

Comment (mkarolin) [reply browser/extensions/api/brave_rewards_api.cc:941]: Right, this is what I did in https://github.com/brave/brave-core/pull/13498/commits/5c71e38347a7a45f7e50d3b868716aac90f7a846#diff-710cdcbdcd014b58902a237fb17ec8e0e5594ce22a58ca182df1e35d68f637bdR232 and its counterpart in https://github.com/brave/brave-core/pull/13498/commits/5c71e38347a7a45f7e50d3b868716aac90f7a846#diff-710cdcbdcd014b58902a237fb17ec8e0e5594ce22a58ca182df1e35d68f637bdR190-R191, but in here these params are being sent to JS, which, from what I understand, doesn't have an equivalent of int64_t. We'd also have to adjust the json definition files to indicate that his value would become a string instead of an integer.

File: ios/app/brave_main_delegate.mm
Comment (mkarolin) [inline ios/app/brave_main_delegate.mm:55]: @kylehickinson @Brandon-T: the above switch has been removed from Chromium. I added a patch/overrides in `[iOS] WIP: disable-enterprise-policy flag was removed.` commit (currently https://github.com/brave/brave-core/pull/13498/commits/bba65256fcdc48fac5516d496f74ec619fce04c4) that look to me like they would be sufficient, but I can't verify. Would be great if you could check at some point.

Comment (kylehickinson) [reply ios/app/brave_main_delegate.mm:55]: @mkarolin Anything we need to check for in particular aside from no crashes at launch/setup?

Comment (mkarolin) [reply ios/app/brave_main_delegate.mm:55]: @kylehickinson looks like originally the switch was added for sync to work (in https://github.com/brave/brave-core/pull/6722), so probably some sync-related activities 😄

Comment (kylehickinson) [reply ios/app/brave_main_delegate.mm:55]: Tested in isolation using `1.42.26`/`104.0.5112.20`:

IMG_5377

Seems like everything's OK. cc @Brandon-T to verify the sync internals screenshot

File: components/brave_rewards/browser/rewards_notification_service_impl.cc
Comment (zenparsing) [inline components/brave_rewards/browser/rewards_notification_service_impl.cc:190]: Will this work for the upgrade case, when "timestamp" will be an int value? It seems like `base::ValueToInt64` will return `nullopt` and we'll get a zero.

One option might be to parse and store as a double instead, and then do the `unit64_t` <-> `double` conversions for the notification structure.

File: browser/extensions/api/brave_rewards_api.cc
Comment (zenparsing) [reply browser/extensions/api/brave_rewards_api.cc:1030]: Apologies for not seeing this before. I see that we had a lot of implicit `uint64_t` to `int` conversions happening and you've made them into explicit casts to `int`.

I think the `uint64_t` conversions to `int` are undesirable. Since JS doesn't know about ints (only doubles), I think it might make the most sense to set all these values as doubles instead of int.

**Review by diracdeltas — APPROVED** [review]
unblocking package.json

**Review by rillian — APPROVED** [review]
r+ for rust-deps

File: components/brave_rewards/browser/rewards_notification_service_impl.cc
Comment (cdesouza-chromium) [reply components/brave_rewards/browser/rewards_notification_service_impl.cc:190]: Another option would be to use `base::ValueToTime`, which would store the timestamp in a human readable format.

File: browser/extensions/api/brave_rewards_api.cc
Comment (mkarolin) [reply browser/extensions/api/brave_rewards_api.cc:1035]: @zenparsing I think that's a great idea. It would be better to solve these outside the scope of the Chromium bump.

File: components/brave_rewards/browser/rewards_notification_service_impl.cc
Comment (mkarolin) [reply components/brave_rewards/browser/rewards_notification_service_impl.cc:190]: yes, I think this will cause problems on "upgrade". Probably best to cast to int for now like we used to do implicitly via SetInteger in StoreRewardsNotifications and leave this line as it was. Then this can be addressed with a follow up issue as you suggested in https://github.com/brave/brave-core/pull/13498#discussion_r918379403

File: browser/translate/brave_translate_browsertest.cc
Comment (iefremov) [inline browser/translate/brave_translate_browsertest.cc:263]: @atuchin-m ptal

Comment (atuchin-m) [reply browser/translate/brave_translate_browsertest.cc:263]: Looks good.
All the tests check the bubble on the active WebContents, so the replacement should be legal.

File: browser/extensions/api/brave_rewards_api.cc
Comment (mkarolin) [reply browser/extensions/api/brave_rewards_api.cc:1035]: Created https://github.com/brave/brave-browser/issues/24015

File: components/brave_rewards/browser/rewards_notification_service_impl.cc
Comment (mkarolin) [reply components/brave_rewards/browser/rewards_notification_service_impl.cc]: Created https://github.com/brave/brave-browser/issues/24015 and reverted this to explicit conversion to int in 1b65ee45ed508f90ffc0bf14de436e483ab88e3f

File: build/rust/Cargo.lock
Comment (mkarolin) [reply build/rust/Cargo.lock:566]: Yes, this is due to 

<pre>
https://chromium.googlesource.com/chromium/src/+/4e69cab0e32be6b6876fb40c18f801822221ceb5

    commit 4e69cab0e32be6b6876fb40c18f801822221ceb5
    Author: Collin Baker <collinbaker@chromium.org>
    Date:   Thu Jun 9 15:40:57 2022 +0000

        Roll autocxx, associated tools, and dependencies

        Bug: 1245714
</pre>

**Review by kylehickinson — APPROVED** [review]
iOS++

**Review by zenparsing — APPROVED** [review]
Rewards ✔️

**Review by tmancey — APPROVED** [review]
+1

File: chromium_src/chrome/browser/ui/views/translate/translate_bubble_controller.cc
Comment (atuchin-m) [inline chromium_src/chrome/browser/ui/views/translate/translate_bubble_controller.cc:11]: Can we make this static method be a free function? ctors of `BraveTranslateBubbleView` & `TranslateBubbleView` are public.

Non blocking note, this can be changed after merging this PR.

**Review by atuchin-m — APPROVED** [review]
translate files +1

File: chromium_src/chrome/browser/ui/views/translate/translate_bubble_controller.cc
Comment (mkarolin) [reply chromium_src/chrome/browser/ui/views/translate/translate_bubble_controller.cc]: Done in [a5e9221](https://github.com/brave/brave-core/pull/13498/commits/a5e92217fb8ea5c511c7a49ee7f290da77e08656)

File: chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc
Comment (bridiver) [inline chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc:37]: why don't we just get rid of all of these changes and overrides to pass ScriptState and just override `probe::ApplyHardwareConcurrencyOverride` which already receives all the information we need?

Comment (bridiver) [reply chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc:37]: `#define ApplyHardwareConcurrencyOverride ApplyBraveHardwareConcurrencyOverride(GetExecutionContext(), hardware_concurrency) probe::ApplyHardwareConcurrencyOverride`

File: patches/chrome-test-BUILD.gn.patch
Comment (bridiver) [inline patches/chrome-test-BUILD.gn.patch:17]: why do we need different conditions here than upstream?

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/prefix_iterator.h
Comment (bridiver) [inline vendor/bat-native-ledger/src/bat/ledger/internal/publisher/prefix_iterator.h:18]: @zenparsing could we use base::BufferIterator here instead?

File: patches/chrome-common-BUILD.gn.patch
Comment (bridiver) [inline patches/chrome-common-BUILD.gn.patch:17]: what is needed from this in skus browser tests?

File: patches/ios-chrome-browser-browser_state-chrome_browser_state_impl.mm.patch
Comment (bridiver) [inline patches/ios-chrome-browser-browser_state-chrome_browser_state_impl.mm.patch:9]: this patch is not necessary, you can do something like `#define BuildBrowserStatePolicyConnector(ARG1, ARG2, ARG3) nullptr; }`

Comment (bridiver) [reply patches/ios-chrome-browser-browser_state-chrome_browser_state_impl.mm.patch:9]: also just for future reference the naming convention for this is just BRAVE_CHROME_BROWSER_STATE_IMPL. Naming it with a specific purpose (CLOSE_IF) goes against the idea of trying to make extensible patches

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/prefix_iterator.h
Comment (bridiver) [reply vendor/bat-native-ledger/src/bat/ledger/internal/publisher/prefix_iterator.h:18]: for follow-up

File: patches/chrome-common-BUILD.gn.patch
Comment (bridiver) [reply patches/chrome-common-BUILD.gn.patch:17]: in any case it should just depend on `//chrome/common` and remove this patch

File: patches/chrome-test-BUILD.gn.patch
Comment (bridiver) [reply patches/chrome-test-BUILD.gn.patch:17]: also if the only purpose of this change is to use this config in test/BUILD.gn then we should just create our own config for it instead of patching this one

File: components/body_sniffer/body_sniffer_url_loader.h
Comment (mkarolin) [reply components/body_sniffer/body_sniffer_url_loader.h:129]: @iefremov @ShivanKaul just pinging you guys here to make sure these changes look ok.

File: chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc]: Addressed in [5e1dfa1](https://github.com/brave/brave-core/pull/13498/commits/5e1dfa1d52180684969c10f204b26b9e5c0681b0) (cc: @pilgrim-brave)

File: patches/chrome-test-BUILD.gn.patch
Comment (mkarolin) [reply patches/chrome-test-BUILD.gn.patch:17]: @cdesouza-chromium ☝️ was this needed because building of the Chromium tests was failing on CI?

File: patches/chrome-common-BUILD.gn.patch
Comment (mkarolin) [reply patches/chrome-common-BUILD.gn.patch]: Fixed in [4e8e261](https://github.com/brave/brave-core/pull/13498/commits/4e8e261cef0726e1ea770188748f4778071fcf31)

**Review by bbondy — APPROVED** [review]
wallet++

File: patches/chrome-test-BUILD.gn.patch
Comment (cdesouza-chromium) [reply patches/chrome-test-BUILD.gn.patch]: Fixed e65909ae8ddb5709ec77b60aeba43594fb4c70dd, according to the suggestion in DMs. I'm also removing the config in our targets because as I remember this was only happening in the upstream tests.

File: chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc
Comment (goodov) [inline chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc:22]: should be either static/in anonymous ns or located right in the function.

**Review by goodov — APPROVED** [review]
lgtm with few nits

File: patches/third_party-blink-public-mojom-permissions_policy-permissions_policy_feature.mojom.patch
Comment (mkarolin) [reply patches/third_party-blink-public-mojom-permissions_policy-permissions_policy_feature.mojom.patch]: Replaced in [4265b53](https://github.com/brave/brave-core/pull/13498/commits/4265b534bb16fa1aed70a180e5e88ef7b2ebcf92)

File: chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc
Comment (mkarolin) [reply chromium_src/third_party/blink/renderer/core/execution_context/navigator_base.cc]: Moved to the function in [eedc83c](https://github.com/brave/brave-core/pull/13498/commits/eedc83c44654f38ebfaf1e60ac1d018e5b8e2ca3)

File: patches/chrome-browser-flags-android-chrome_feature_list.cc.patch
Comment (mkarolin) [reply patches/chrome-browser-flags-android-chrome_feature_list.cc.patch]: Done in [034234e](https://github.com/brave/brave-core/pull/13498/commits/034234e7f98015850e90328e52cbd38ffc6b0082), haven't tested locally, though. Will wait for the CI. cc: @samartnik

File: components/body_sniffer/body_sniffer_url_loader.h
Comment (ShivanKaul) [reply components/body_sniffer/body_sniffer_url_loader.h:129]: So `OnReceiveResponse` and `OnStartLoadingResponseBody` were combined into one, and we need the next_body_consumer_handle_ to keep track of whether it's De-AMP or Speedreader consuming the body?

Comment (mkarolin) [reply components/body_sniffer/body_sniffer_url_loader.h:129]: We need to know the consuming handle for the body ahead of time because when the throttles are chained like with De-AMP and Speedreader, if we don't supply the handle then the other throttle(s) in the chain will never receive the body.

**Review by ShivanKaul — APPROVED** [review]

File: components/body_sniffer/body_sniffer_url_loader.h
Comment (iefremov) [reply components/body_sniffer/body_sniffer_url_loader.h:129]: thanks @mkarolin

**Comment by mkarolin** [comment]
@tmancey, PTAL at [a00cc7e](https://github.com/brave/brave-core/pull/13498/commits/a00cc7e93331c854b3fcec9b8c004399b3a3f5c6) - not sure why it isn't giving an error on master.

File: patches/third_party-polymer-v3_0-BUILD.gn.patch
Comment (mkarolin) [inline patches/third_party-polymer-v3_0-BUILD.gn.patch:10]: @petemill Is there a better way to do this? (commit: https://github.com/brave/brave-core/pull/13498/commits/109b03ab5e6674329b3620bb0cecfa4c9cfceefe). Stems from https://github.com/brave/brave-core/pull/14067

**Pull request merged**
